File: plugin/src/test/java/com/jtprince/bingo/plugin/automarking/itemtrigger/TestItemTriggerYaml.java
package com.jtprince.bingo.plugin.automarking.itemtrigger;

import org.junit.jupiter.api.Test;

import java.util.HashMap;
import java.util.Map;

import static org.junit.jupiter.api.Assertions.*;
import static org.junit.jupiter.api.Assertions.assertEquals;

/**
 * Test the functionality of the Item Trigger YAML parser.
 */
public class TestItemTriggerYaml {
    final static String YAML_PATH = "/test_item_triggers.yml";
    final ItemTriggerYaml yaml;
    final Map<String, Integer> testVariables = new HashMap<>();

    public TestItemTriggerYaml() {
        yaml = ItemTriggerYaml.fromFile(getClass().getResourceAsStream(YAML_PATH));
    }

    @Test
    void testYamlBasic() {
        ItemTriggerYaml.MatchGroup g = yaml.get("jm_book_quill");
        assertNotNull(g);
        assertTrue(g.nameMatches("minecraft:writable_book"));
        assertFalse(g.nameMatches("minecraft:book"));
        assertEquals(1, g.getUnique(testVariables));
        assertEquals(1, g.getTotal(testVariables));
        assertEquals(0, g.children.size());
    }

    @Test
    void testYamlNonexistent() {
        ItemTriggerYaml.MatchGroup g = yaml.get("nonexistent_goal_id");
        assertNull(g);
    }

    @Test
    void testYamlGroups() {
        ItemTriggerYaml.MatchGroup g = yaml.get("jm_different_edible");
        assertNotNull(g);
        assertTrue(g.nameMatches("minecraft:beetroot"));
        assertTrue(g.nameMatches("minecraft:regex_stew"));
        assertFalse(g.nameMatches("minecraft:potato"));  // In child
        assertEquals(6, g.getUnique(testVariables));
        assertEquals(2, g.children.size());

        ItemTriggerYaml.MatchGroup potatoChild = g.children.get(0);
        assertNotNull(potatoChild);
        assertTrue(potatoChild.nameMatches("minecraft:potato"));
        assertFalse(potatoChild.nameMatches("minecraft:beetroot"));
        assertEquals(1, potatoChild.getUnique(testVariables));
        assertEquals(1, potatoChild.getTotal(testVariables));
        assertEquals(0, potatoChild.children.size());
    }
}

File: kotlinplugin/src/test/kotlin/com/jtprince/bingo/kplugin/automark/TestItemTriggerYaml.kt
package com.jtprince.bingo.kplugin.automark

import org.junit.jupiter.api.Test
import kotlin.test.*

/**
 * Test the functionality of the Item Trigger YAML parser.
 */
class TestItemTriggerYaml {
    private val yaml = ItemTriggerYaml.fromFile(javaClass.getResourceAsStream("/test_item_triggers.yml"))
    private val noVariables = mapOf<String, Int>()

    @Test
    fun `A simple goal with no variables can be parsed`() {
        val matchGroup = yaml["jm_book_quill"]
        assertNotNull(matchGroup)
        assertTrue(matchGroup.nameMatches("minecraft:writable_book"))
        assertFalse(matchGroup.nameMatches("minecraft:book"))
        assertEquals(1, matchGroup.unique(noVariables))
        assertEquals(1, matchGroup.total(noVariables))
        assertEquals(0, matchGroup.children.size)
    }

    @Test
    fun `A match group that does not exist returns null`() {
        val matchGroup = yaml["nonexistent_goal_id"]
        assertNull(matchGroup)
    }

    @Test
    fun `A match group with children can be parsed`() {
        val matchGroup = yaml["jm_different_edible"]
        assertNotNull(matchGroup)
        assertTrue(matchGroup.nameMatches("minecraft:beetroot"))
        assertTrue(matchGroup.nameMatches("minecraft:regex_stew"))
        assertFalse(matchGroup.nameMatches("minecraft:potato")) // In child
        assertEquals(6, matchGroup.unique(noVariables))
        assertEquals(2, matchGroup.children.size)
        val potatoChild = matchGroup.children[0]
        assertNotNull(potatoChild)
        assertTrue(potatoChild.nameMatches("minecraft:potato"))
        assertFalse(potatoChild.nameMatches("minecraft:beetroot"))
        assertEquals(1, potatoChild.unique(noVariables))
        assertEquals(1, potatoChild.total(noVariables))
        assertEquals(0, potatoChild.children.size)
    }
}
